packages/orderbook/src/order_provider/sra_websocket_order_provider.ts

```diff
@@ -0,0 +1,182 @@
+
+import { BaseSRAOrderProvider } from './base_sra_order_provider';
+
+const RECORD_COUNT = 100;
+
+export class SRAWebsocketOrderProvider extends BaseSRAOrderProvider {
+    private readonly _websocketEndpoint: string;
+    private readonly _wsSubscriptions: Map<string, OrdersChannelSubscriptionOpts> = new Map();
+    private _ordersChannel?: OrdersChannel;
+    private _isDestroyed = false;
+    private _isConnecting = false;
+
+    /**
+     * Instantiates a HTTP and WS [Standard Relayer API](https://github.com/0xProject/standard-relayer-api) Order Provider
+     * @param opts `SRAWebsocketOrderProviderOpts` containing the websocketEndpoint and the httpEndpoint to an SRA backend.
+     * @param orderStore The `OrderStore` where orders are added and removed from
+     */
+    constructor(opts: SRAWebsocketOrderProviderOpts, orderStore: OrderStore) {
+        super(orderStore, opts.httpEndpoint, RECORD_COUNT, opts.networkId);
+        assert.isUri('websocketEndpoint', opts.websocketEndpoint);
+        this._websocketEndpoint = opts.websocketEndpoint;
+    }
+
+    /**
+     * Creates a websocket subscription and fetches the current orders from SRA. If a websocket
+     * connection already exists this function is a noop.
+     * @param makerAssetData the Maker Asset Data
+     * @param takerAssetData the Taker Asset Data
+     */
+    public async createSubscriptionForAssetPairAsync(makerAssetData: string, takerAssetData: string): Promise<void> {
+        // If we've previously been destroyed then reset
+        this._isDestroyed = false;
+        const assetPairKey = OrderStore.getKeyForAssetPair(makerAssetData, takerAssetData);
+        if (this._wsSubscriptions.has(assetPairKey)) {
+            return;
+        }
+        return this._fetchAndCreateSubscriptionAsync(makerAssetData, takerAssetData);
+    }
+
+    /**
+     * Destroys the order provider, removing any subscriptions
+     */
+    public async destroyAsync(): Promise<void> {
+        this._isDestroyed = true;
+        this._wsSubscriptions.clear();
+        if (this._ordersChannel) {
+            this._ordersChannel.close();
+            this._ordersChannel = undefined;
+        }
+    }
+
+    /**
+     * Creates a websocket subscription. If the inital websocket connnection
+     * does not exist, it is created.
+     * @param makerAssetData the Maker Asset Data
+     * @param takerAssetData the Taker Asset Data
+     */
+    private async _createWebsocketSubscriptionAsync(makerAssetData: string, takerAssetData: string): Promise<void> {
+        // Prevent creating multiple channels
+        while (this._isConnecting && !this._ordersChannel) {
+            await utils.delayAsync(100);
+        }
+        if (!this._ordersChannel) {
+            this._isConnecting = true;
+            try {
+                this._ordersChannel = await this._createOrdersChannelAsync();
+            } finally {
+                this._isConnecting = false;
+            }
```

Comment:
Don't we need a `catch` clause here? Otherwise it'll continue without an active subscription.

Reply:
since there is no catch it will throw the underlying error whilst also updating `_isConnecting` to false. So we don't need to catch and re-throw.